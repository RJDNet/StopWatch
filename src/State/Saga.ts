import {
    default as createSagaMiddleware,
    SagaMiddleware
} from 'redux-saga';
import { all } from 'redux-saga/effects';

import stopWatchContainerRootSaga from '../StopWatchContainer/Sagas/RootSaga';
import stopWatchRootSaga from '../StopWatch/Sagas/RootSaga';

export const sagaMiddleware: SagaMiddleware<{}> = createSagaMiddleware();

export function startRootSaga(): void {
    sagaMiddleware.run(rootSaga);
}

export function* rootSaga() {
    yield all([
        stopWatchContainerRootSaga(),
        stopWatchRootSaga()
    ]);
}
