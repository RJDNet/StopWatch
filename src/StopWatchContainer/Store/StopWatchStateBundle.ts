import { IRemoveStopWatch } from './StopWatchStateBundle';
import uuidv1 from 'uuid/v1';

// Action Types
export enum ActionTypes {
    ADD_STOPWATCH = 'ADD_STOPWATCH',
    ADDX_STOPWATCH = 'ADDX_STOPWATCH',
    REMOVEALL_STOPWATCH = 'REMOVEALL_STOPWATCH',
    REMOVE_STOPWATCH = 'REMOVE_STOPWATCH',
    START_STOPWATCH = 'START_STOPWATCH',
    STOP_STOPWATCH = 'STOP_STOPWATCH',
    TICK_STOPWATCH = 'TICK_STOPWATCH',
    RESET_STOPWATCH = 'RESET_STOPWATCH'
}

// Action Interfaces
export interface IAddStopWatch {
    type: ActionTypes.ADD_STOPWATCH;
}

export interface IAddXStopWatch {
    type: ActionTypes.ADDX_STOPWATCH;
    payload: {
        amount: number;
    }
}

export interface IRemoveAllStopWatch {
    type: ActionTypes.REMOVEALL_STOPWATCH;
}

export interface IRemoveStopWatch {
    type: ActionTypes.REMOVE_STOPWATCH;
    payload: {
        key: string;
    }
}

export interface IStartStopWatch {
    type: ActionTypes.START_STOPWATCH;
    payload: {
        key: string;
    }
}

export interface IStopStopWatch {
    type: ActionTypes.STOP_STOPWATCH;
    payload: {
        key: string;
    }
}

export interface ITickStopWatch {
    type: ActionTypes.TICK_STOPWATCH;
    payload: {
        key: string;
    }
}

export interface IResetStopWatch {
    type: ActionTypes.RESET_STOPWATCH;
    payload: {
        key: string;
    }
}

// Consolidate Action Interfaces
export type Actions = IAddStopWatch |
    IAddXStopWatch |
    IRemoveAllStopWatch |
    IRemoveStopWatch |
    IStartStopWatch |
    IStopStopWatch |
    ITickStopWatch |
    IResetStopWatch;

// Action Creators
export const actionCreators = {
    addWatch(): IAddStopWatch {
        const action: IAddStopWatch = {
            type: ActionTypes.ADD_STOPWATCH
        }
        
        return action;
    },
    addXWatch(amount: number): IAddXStopWatch {
        const action: IAddXStopWatch = {
            type: ActionTypes.ADDX_STOPWATCH,
            payload: {
                amount
            }
        }
        
        return action;
    },
    removeAllWatches(): IRemoveAllStopWatch {
        const action: IRemoveAllStopWatch = {
            type: ActionTypes.REMOVEALL_STOPWATCH
        }
        
        return action;
    },
    removeWatch(key: string): IRemoveStopWatch {
        const action: IRemoveStopWatch = {
            type: ActionTypes.REMOVE_STOPWATCH,
            payload: {
                key
            }
        }
        
        return action;
    },
    startWatch(key: string): IStartStopWatch {
        const action: IStartStopWatch = {
            type: ActionTypes.START_STOPWATCH,
            payload: {
                key
            }
        }
        
        return action;
    },
    stopWatch(key: string): IStopStopWatch {
        const action: IStopStopWatch = {
            type: ActionTypes.STOP_STOPWATCH,
            payload: {
                key
            }
        }
            
        return action;
    },
    tickWatch(key: string): ITickStopWatch {
        const action: ITickStopWatch = {
            type: ActionTypes.TICK_STOPWATCH,
            payload: {
                key
            }
        }
            
        return action;
    },
    resetWatch(key: string): IResetStopWatch {
        const action: IResetStopWatch = {
            type: ActionTypes.RESET_STOPWATCH,
            payload: {
                key
            }
        }
        
        return action;
    }
}

export interface IStopWatch {
    time: number;
    running: boolean;
    key: string;
}

export interface IStopWatchMap {
    [key: string]: IStopWatch;
}

export interface IStopWatchState {
    stopWatches: IStopWatchMap;
    stopWatchList: string[];
}

// Default value of store
export function getInitialStopWatchState(): Readonly<IStopWatchState> {
    const state: Readonly<IStopWatchState> = {
        stopWatches: {},
        stopWatchList: []
    }

    return state;
}

// Sub Reducers
export function addWatch(state: IStopWatchState): IStopWatchState {
    const key: string = uuidv1();

    const update: IStopWatchState = {
        ...state,
        stopWatches: {
            ...state.stopWatches,
            [key]: {
                time: 0,
                running: false,
                key
            }
        },
        stopWatchList: [...state.stopWatchList, key]
    }

    return update;
}

export function addXWatch(state: IStopWatchState, action: IAddXStopWatch): IStopWatchState {
    const { amount } = action.payload;
    let update = state;
    
    for(let i = 1; i <= amount; i++) {
        const key: string = uuidv1();
        let newUpdate: IStopWatchState = {
            ...update,
            stopWatches: {
                ...update.stopWatches,
                [key]: {
                    time: 0,
                    running: false,
                    key
                }
            },
            stopWatchList: [...update.stopWatchList, key]
        }

        update = newUpdate;
    }
    

    return update;
}

function removeAllWatches(state: IStopWatchState): IStopWatchState {
    const update: IStopWatchState = {
        ...state,
        stopWatches: {},
        stopWatchList: []
    }

    return update;
}

function removeWatch(state: IStopWatchState, action: IRemoveStopWatch): IStopWatchState {
    const { key } = action.payload;
    const delKey = {
        ...state
    }
    delete delKey.stopWatches[key]

    const update: IStopWatchState = {
        ...state,
        stopWatches: {
            ...delKey.stopWatches
        },
        stopWatchList: state.stopWatchList.filter(f => f !== key)
    }

    return update;
}

function startWatch(state: IStopWatchState, action: IStartStopWatch): IStopWatchState {
    const { key } = action.payload;

    const update: IStopWatchState = {
        ...state,
        stopWatches: {
            ...state.stopWatches,
            [key]: {
                ...state.stopWatches[key],
                running: true
            }
        }
    }

    return update;
}

function stopWatch(state: IStopWatchState, action: IStopStopWatch): IStopWatchState {
    const { key } = action.payload;

    const update: IStopWatchState = {
        ...state,
        stopWatches: {
            ...state.stopWatches,
            [key]: {
                ...state.stopWatches[key],
                running: false
            }
        }
    }

    return update;
}

function tickWatch(state: IStopWatchState, action: ITickStopWatch): IStopWatchState {
    const { key } = action.payload;

    if (state.stopWatches[key] && state.stopWatches[key].running) {
        const time: number = state.stopWatches[key].time + 1;
        const update: IStopWatchState = {
            ...state,
            stopWatches: {
                ...state.stopWatches,
                [key]: {
                    ...state.stopWatches[key],
                    time
                }
            }
        }
    
        return update;
    }

    return state;
}

function resetWatch(state: IStopWatchState, action: IResetStopWatch): IStopWatchState {
    const { key } = action.payload;

    const update: IStopWatchState = {
        ...state,
        stopWatches: {
            ...state.stopWatches,
            [key]: {
                ...state.stopWatches[key],
                running: false,
                time: 0
            }
        }
    }

    return update;
}

// Reducer
export const stopWatchReducer = (state: IStopWatchState = getInitialStopWatchState(), action: Actions ) => {
    switch (action.type) {
      case ActionTypes.ADD_STOPWATCH:
        return addWatch(state);
      case ActionTypes.ADDX_STOPWATCH:
        return addXWatch(state, action);
      case ActionTypes.REMOVEALL_STOPWATCH:
        return removeAllWatches(state);
      case ActionTypes.REMOVE_STOPWATCH:
        return removeWatch(state, action);
      case ActionTypes.START_STOPWATCH: 
        return startWatch(state, action);
      case ActionTypes.STOP_STOPWATCH: 
        return stopWatch(state, action);
      case ActionTypes.TICK_STOPWATCH: 
        return tickWatch(state, action);
      case ActionTypes.RESET_STOPWATCH: 
        return resetWatch(state, action);
      default: return state;
    }
}

// Helpers
// function uuidv4() {
//     return 'xxxxxxxx-xxxx-4xxx-yxxx-xxxxxxxxxxxx'.replace(/[xy]/g, function(c) {
//       var r = Math.random() * 16 | 0, v = c == 'x' ? r : (r & 0x3 | 0x8);
//       return v.toString(16);
//     });
//   }

// export function uuidv4() {
//     var dt = new Date().getTime();
//     var uuid = 'xxxxxxxx-xxxx-4xxx-yxxx-xxxxxxxxxxxx'.replace(/[xy]/g, function(c) {
//         var r = (dt + Math.random()*16)%16 | 0;
//         dt = Math.floor(dt/16);
//         return (c==='x' ? r :(r&0x3|0x8)).toString(16);
//     });
//     return uuid;
// }
