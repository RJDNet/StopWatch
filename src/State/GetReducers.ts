import { stopWatchReducer } from '../StopWatchContainer/Store/StopWatchStateBundle';

export function getReducers(): {} {
    return {
        stopWatchState: stopWatchReducer
    };
}