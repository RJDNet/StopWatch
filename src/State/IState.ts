import { IStopWatchState } from '../StopWatchContainer/Store/StopWatchStateBundle';

export interface IState {
    stopWatchState: IStopWatchState;
}