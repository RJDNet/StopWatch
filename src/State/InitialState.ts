import {
    IStopWatchState,
    getInitialStopWatchState
} from '../StopWatchContainer/Store/StopWatchStateBundle';

import { IState } from './IState';

export function initialStoreState(): Readonly<IState> {
    const stopWatchState: IStopWatchState = getInitialStopWatchState();

    return {
        stopWatchState: stopWatchState
    };
}