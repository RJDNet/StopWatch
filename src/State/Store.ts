import {
    Store,
    Unsubscribe
} from 'redux';

import { IState } from './IState';

export let store: Readonly<Store<IState>>;

export function getState(): Readonly<IState> {
    return store.getState();
}

export function setStore(storeToSet: Readonly<Store<IState>>): void {
    store = storeToSet;
}

export function dispatch(action: { type: string }): void {
    if (store !== undefined) {
        store.dispatch(action);
    }
}

export function subscribe(listener: () => void): Unsubscribe {
    return store.subscribe(listener);
}