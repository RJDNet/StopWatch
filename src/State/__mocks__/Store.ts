import { 
    createStore, 
    Store, 
    combineReducers,
    Reducer
  } from 'redux';

import { IState } from '../IState';
import {
    IStopWatchState,
    getInitialStopWatchState
} from '../../StopWatchContainer/Store/StopWatchStateBundle';
import { getReducers } from '../../State/GetReducers';

export function initialStoreState(): Readonly<IState> {
    const stopWatchState: IStopWatchState = { ...getInitialStopWatchState() };

    return {
        stopWatchState
    };
}

export function getState(): Readonly<IState> {
    const state: Readonly<IState> = initialStoreState();

    return { 
        ...state 
    };
}

export const dispatch = jest.fn();
export const subscribe = jest.fn();

export function setupStore(initialState: IState = getState()): Readonly<Store<IState>> {
    const reducer: Reducer<{}> = combineReducers(getReducers());

    return createStore(
        reducer,
        initialState,
        window.__REDUX_DEVTOOLS_EXTENSION__ && window.__REDUX_DEVTOOLS_EXTENSION__()
    ) as Store<IState>;
}
  
export function getMockState(): Readonly<IState> {
    return store.getState();
}

export let store: any = setupStore();

function setStore(storeToSet: Readonly<Store<IState>>): void {
    store = storeToSet;
}

setStore(store);
getMockState();
  