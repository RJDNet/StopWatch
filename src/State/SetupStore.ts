import { composeWithDevTools } from 'redux-devtools-extension';
import {
  createStore,
  applyMiddleware,
  combineReducers,
  Store,
  Reducer
} from 'redux';
import { sagaMiddleware } from './Saga';
import { IState } from './IState';
import { initialStoreState } from './InitialState';
import { getReducers } from './GetReducers';

declare global {
    interface Window {
      __REDUX_DEVTOOLS_EXTENSION__: Function
    }
}

const composeEnhancers: <GenericStoreEnhancer>(f: GenericStoreEnhancer) => GenericStoreEnhancer = composeWithDevTools({
  serialize: true
});

export function setupStore(initialState: IState = initialStoreState()): Readonly<Store<IState>> {
  const reducer: Reducer<{}> = combineReducers(getReducers());
  return createStore(
      reducer,
      initialState,
      composeEnhancers(applyMiddleware(sagaMiddleware))
  ) as Store<IState>;
}
