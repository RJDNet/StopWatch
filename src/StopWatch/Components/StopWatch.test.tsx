import React from 'react';
import { Provider } from "react-redux";
import { 
  shallow, 
  ShallowWrapper 
} from 'enzyme';
import toJson from 'enzyme-to-json';
import { v1 as uuidv1 } from 'uuid';

import { 
  store, 
  getState 
} from '../../State/__mocks__/Store';
import { IState } from '../../State/IState';
import { ActionTypes } from '../../StopWatchContainer/Store/StopWatchStateBundle';
import ConnectedStopWatch, { 
  StopWatch, 
  mapStateToProps,
  IStopWatchProps, 
  IMapProps
} from './StopWatch';

describe('StopWatch Component', () => {
  const dispatch = jest.fn();
  const key: string = uuidv1();

  const props: IStopWatchProps = {
    className: 'stopwatch',
    id: 'testerid',
    runningFilter: 'All',
    stopWatch: {
      time: 0,
      running: false,
      key: key,
    },
    dispatch
  }

  test('StopWatch renders correctly with all props & no redux store', () => {
    const tree: ShallowWrapper = shallow(
      <StopWatch {...props} />
    );
      
    expect(toJson(tree)).toMatchSnapshot();
  });

  test('StopWatch renders correctly with className, id * runningFilter props with redux store', () => {
    const tree: ShallowWrapper = shallow(
      <Provider store={store}>
        <ConnectedStopWatch className={'stopwatch'} id={'testerid'} runningFilter={'All'} />
      </Provider>
    );
      
    expect(toJson(tree)).toMatchSnapshot();
  });

  test('mapStateToProps returns expected state', () => {
    const stopWatchState = getState();
    const state: IState = {
      ...stopWatchState,
      stopWatchState: {
        stopWatches: {
          testid: 
            {
              time: 0,
              running: false,
              key: 'testid'
            }
        },    
        stopWatchList: ['testid'],
        filteredList: []
      }
    }
    
    const own: IMapProps = {
      className: 'stopwatch',
      id: 'testid',
      runningFilter: 'All'
    }
    
    const result = mapStateToProps(state, own);

    expect(result).toBeDefined();
    expect(result.stopWatch).toEqual({ time: 0, running: false, key: 'testid'})
  });

  test('test button', () => {
    const tree: ShallowWrapper = shallow(
      <StopWatch {...props} />
    );

    const selectElement = tree.find('div').hasClass('stopwatch');
    expect(selectElement).toBe(true);
  });

  test('button click should startstop stopwatch', () => {
    const tree: ShallowWrapper = shallow(
      <StopWatch {...props} />
    );

    tree
      .find('button.stopwatch__startstop-button')
      .simulate('click');
    expect(dispatch).toHaveBeenCalled();
    expect(dispatch).toBeCalledWith({
      type: ActionTypes.START_STOPWATCH,
      payload: {
        key: 'testerid'
      }
    });
  });

  test('button click should reset stopwatch', () => {
    const tree: ShallowWrapper = shallow(
      <StopWatch {...props} />
    );

    tree
      .find('button.stopwatch__reset-button')
      .simulate('click');
    expect(dispatch).toHaveBeenCalled();
    expect(dispatch).toBeCalledWith({
      type: ActionTypes.RESET_STOPWATCH,
      payload: {
        key: 'testerid'
      }
    });
  });

  test('button click should remove stopwatch', () => {
    const tree: ShallowWrapper = shallow(
      <StopWatch {...props} />
    );

    tree
      .find('button.stopwatch__remove-button')
      .simulate('click');
    expect(dispatch).toHaveBeenCalled();
    expect(dispatch).toBeCalledWith({
      type: ActionTypes.REMOVE_STOPWATCH,
      payload: {
        key: 'testerid'
      }
    });
  });
});