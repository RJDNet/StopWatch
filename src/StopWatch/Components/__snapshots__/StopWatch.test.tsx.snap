// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`StopWatch Component StopWatch renders correctly with all props & no redux store 1`] = `
<div
  className="stopwatch stopwatch"
>
  <p
    className="stopwatch__timer"
  >
    0
    :
    00
  </p>
  <button
    className="stopwatch__startstop-button stopwatch__startstop-button--start"
    onClick={[Function]}
  >
    START
  </button>
  <button
    className="stopwatch__reset-button"
    onClick={[Function]}
  >
    RESET
  </button>
  <button
    className="stopwatch__remove-button"
    onClick={[Function]}
  >
    REMOVE
  </button>
</div>
`;

exports[`StopWatch Component StopWatch renders correctly with className, id * runningFilter props with redux store 1`] = `
<ContextProvider
  value={
    Object {
      "store": Object {
        "dispatch": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
        Symbol(observable): [Function],
      },
      "subscription": Subscription {
        "handleChangeWrapper": [Function],
        "listeners": Object {
          "clear": [Function],
          "get": [Function],
          "notify": [Function],
          "subscribe": [Function],
        },
        "onStateChange": [Function],
        "parentSub": undefined,
        "store": Object {
          "dispatch": [Function],
          "getState": [Function],
          "replaceReducer": [Function],
          "subscribe": [Function],
          Symbol(observable): [Function],
        },
        "unsubscribe": [Function],
      },
    }
  }
>
  <Connect(StopWatch)
    className="stopwatch"
    id="testerid"
    runningFilter="All"
  />
</ContextProvider>
`;
